wasi/wasi_test.go

```diff
@@ -0,0 +1,261 @@
+package wasi
+
+import (
+	"encoding/binary"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/tetratelabs/wazero/wasm"
+)
+
+func TestWasiStringArray(t *testing.T) {
+	tests := []struct {
+		name            string
+		args            []string
+		expectedBufSize uint32
+	}{
+		{
+			name:            "empty",
+			args:            []string{},
+			expectedBufSize: 0,
+		},
+		{
+			name:            "simple",
+			args:            []string{"foo", "bar", "foobar", "", "baz"},
+			expectedBufSize: 20,
+		},
+		{
+			name: "utf-8 string",
+			// "😨", "🤣", and "️🏃‍♀️" have 4, 4, and 13 bytes respectively
+			args:            []string{"😨🤣🏃\u200d♀️", "foo", "bar"},
+			expectedBufSize: 30,
+		},
+		{
+			name:            "invalid utf-8 string",
+			args:            []string{"\xff\xfe\xfd", "foo", "bar"},
+			expectedBufSize: 12,
+		},
+	}
+
+	for _, tt := range tests {
+		tc := tt
+
+		t.Run(tc.name, func(t *testing.T) {
+			wasiStringsArray, err := newWASIStringArray(tc.args)
+			require.NoError(t, err)
+
+			require.Equal(t, tc.expectedBufSize, wasiStringsArray.totalBufSize)
+			require.Equal(t, len(wasiStringsArray.nullTerminatedValues), len(tc.args))
+			for i, arg := range tc.args {
+				wasiString := wasiStringsArray.nullTerminatedValues[i]
+				require.Equal(t, wasiString[0:len(wasiString)-1], []byte(arg))
+				require.Equal(t, wasiString[len(wasiString)-1], byte(0))
+			}
+		})
+	}
+}
+
+func TestArgsAPISucceed(t *testing.T) {
+	tests := []struct {
+		name            string
+		args            []string
+		expectedArgs    [][]byte
+		expectedBufSize uint32
+	}{
+		{
+			name:            "no args",
+			args:            nil,
+			expectedArgs:    [][]byte{},
+			expectedBufSize: 0,
+		},
+		{
+			name:            "empty",
+			args:            []string{},
+			expectedArgs:    [][]byte{},
+			expectedBufSize: 0,
+		},
+		{
+			name: "simple",
+			args: []string{"foo", "bar", "foobar", "", "baz"},
+			expectedArgs: [][]byte{
+				[]byte("foo\x00"),
+				[]byte("bar\x00"),
+				[]byte("foobar\x00"),
+				[]byte("\x00"),
+				[]byte("baz\x00"),
+			},
+			expectedBufSize: 20,
+		},
+		{
+			name: "utf-8 string",
+			// "😨", "🤣", and "️🏃‍♀️" have 4, 4, and 13 bytes respectively
+			args: []string{"😨🤣🏃\u200d♀️", "foo", "bar"},
+			expectedArgs: [][]byte{
+				[]byte("😨🤣🏃\u200d♀️\x00"),
+				[]byte("foo\x00"),
+				[]byte("bar\x00"),
+			},
+			expectedBufSize: 30,
+		},
+		{
+			name: "invalid utf-8 string",
+			args: []string{"\xff\xfe\xfd", "foo", "bar"},
+			expectedArgs: [][]byte{
+				[]byte("\xff\xfe\xfd\x00"),
+				[]byte("foo\x00"),
+				[]byte("bar\x00"),
+			},
+			expectedBufSize: 12,
+		},
+	}
+
+	for _, tt := range tests {
+		tc := tt
+
+		t.Run(tc.name, func(t *testing.T) {
+			opts := []Option{}
+			if tc.args != nil {
```

Comment:
args == nil is correctly handled in `newWASIStringArray` so do we need this branch?

Reply:
This branch is necessary to test the behavior of args API when a `wasiEnvironment` is instantiated without `Args` option in the first place, while the test case of `args == nil` of `newWasiStringArray` tests the behavior of `newWasiStringArray` when nil is explicitly passed to `newWasiStringArray`.